### PR TITLE
Lsn5 junit

### PR DESCRIPTION
**Pull request by CarlyLanglois**
java-web-dev-exercises is updated with the starter code for this section

### CODE REVIEW

File: src/chapters/unit-testing/testing-in-java.rst [hunk 1]
```diff
@@ -0,0 +1,81 @@
+Testing in Java
+===============
+
+.. index:: ! automated testing, ! unit testing, ! JUnit, !JUnit4
+
+You probably know by now the importance of testing your code.
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: Rephrasing suggestion: How about, "You already recognize the importance of..."

Implies existing competence rather than potential ignorance.

File: src/chapters/unit-testing/testing-in-java.rst [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+You probably know by now the importance of testing your code. 
+**Automated testing** in Java, like any language, ensures that 
+your code works as expected, every time it runs. Tests also 
+function as documentation, giving an observer instructions 
+on how to properly execute your classes and methods.
+
+This course covers **unit testing** in Java with a test runner called
+**JUnit4**. Unit testing breaks down the codebase into its smallest
+building blocks, individual statements and methods.
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: "...building blocks, which are statements and methods."
OR
"...building blocks---statements and methods."

I'm not sure that the comma can stand on its own in this case.

File: src/chapters/unit-testing/testing-in-java.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+building blocks, individual statements and methods. 
+
+Why We Test
+^^^^^^^^^^^
+
+Refactoring
+~~~~~~~~~~~
+
+Because they work on the most basic functionality of your code, 
+unit tests safe guard against bugs introduced in refactoring.
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: Typo: "safeguard" is one word.

File: src/chapters/unit-testing/testing-in-java.rst [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+Imagine this common workflow: You practice TDD, writing 
+your tests to stipulate your class's code. You write your code
+to pass your tests. Months later, a stakeholder in your project's
+use requests that you refactor your code using different syntax.
+The features will be the same, but the implementation changes. 
+Unit tests help in this scenario in that, changes to implementation
+should not require changes to outcome. Thus, if your tests continue
+to pass after the refactor, you can move on, knowing you have not 
+inadvertently introduced a bug.
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: How about a set of bullet points (with some minor rewording)?

1) You use TDD to write tests that stipulate your class' behavior.
2) You code your class such that all tests pass.
3) Later, a stakeholder for your project requests that you refactor your code using a different syntax. The features remain the same, but the implementation changes.

Then you can dive into the "outcomes should not change" piece.

For the last sentence, how about, "If your tests still pass after the refactor, you can move on, because you know you have not introduced a bug."

File: src/chapters/unit-testing/testing-in-java.rst [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+Documentation
+~~~~~~~~~~~~~
+
+Unit tests are the most enlightened form of documentation. Again, 
+because they address the most fundamental tasks of your classes,
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst:52]: "Since they address..."

File: src/chapters/unit-testing/testing-in-java.rst [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+unit tests serve as live-code use-cases. You may also have an
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst:53]: "You may also have..." seems like the start of a new paragraph.  This could help introduce breaks between the separate points.

File: src/chapters/unit-testing/testing-in-java.rst [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+inadvertently introduced a bug.
+
+Documentation
+~~~~~~~~~~~~~
+
+Unit tests are the most enlightened form of documentation. Again, 
+because they address the most fundamental tasks of your classes,
+unit tests serve as live-code use-cases. You may also have an 
+external documentation directory with examples of how to run your
+project, or perhaps you hve been writing comments within your code
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: Typo: "hve".

File: src/chapters/unit-testing/testing-in-java.rst [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+to best communicate with your teammates about your changes. Both of
+these are great choices and should be done when possible, but they
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: "...when possible. However, they also require..."

File: src/chapters/unit-testing/testing-in-java.rst [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+Unit tests are the most enlightened form of documentation. Again, 
+because they address the most fundamental tasks of your classes,
+unit tests serve as live-code use-cases. You may also have an 
+external documentation directory with examples of how to run your
+project, or perhaps you hve been writing comments within your code
+to best communicate with your teammates about your changes. Both of
+these are great choices and should be done when possible, but they 
+also both require more forethought to maintain. Each time you update
+your code, you may not always be also updating the documentation and
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: Grammar: "...you might not have a chance to update the documentation..."

File: src/chapters/unit-testing/testing-in-java.rst [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+Unit tests are the most enlightened form of documentation. Again, 
+because they address the most fundamental tasks of your classes,
+unit tests serve as live-code use-cases. You may also have an 
+external documentation directory with examples of how to run your
+project, or perhaps you hve been writing comments within your code
+to best communicate with your teammates about your changes. Both of
+these are great choices and should be done when possible, but they 
+also both require more forethought to maintain. Each time you update
+your code, you may not always be also updating the documentation and 
+the comments. However, with unit testing, you have a more obvious reminder
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: "However" is extraneous here.

Also, how about, "With unit testing, you get a more obvious reminder that a change occurred, especially if a test fails after an update."

File: src/chapters/unit-testing/testing-in-java.rst [hunk 11]
```diff
@@ -0,0 +1,81 @@
+Testing in Java
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst:1]: What do you think about adding a link to the JavaScript unit testing intro page?

This page does essentially the same thing, but unit 1 provides the rationale behind why automatic testing is better than manual testing.  In unit 2, we just assume the students already have some testing motivation. Do you think it would be valuable to provide a link for those who did not complete unit 1?

File: src/chapters/unit-testing/testing-in-java.rst [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+Below are some best practices to keep in mind when writing unit tests, in any language.
+
+#. The AAAs
+
+   The AAAs of of unit testing refers to the pattern to follow when 
+   writing your unit tests. 
+
+   #. Arrange the variables your test requires
+   #. Act on the methods your test requires
+   #. Assert the anticipated comparison of the expected and the actual value
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: To get lettered sub-bullets in a non-concept check admonition, we actually have to type the letters.

Also, how about emphasizing each "A" word (e.g. *Arrange*).

Typo: "...expected and actual values"

File: src/chapters/unit-testing/testing-in-java.rst [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+your code, you may not always be also updating the documentation and 
+the comments. However, with unit testing, you have a more obvious reminder
+that a change has been made if a test fails after a code change.
+
+.. _testing-best-practices:
+
+Testing Best Practices
+^^^^^^^^^^^^^^^^^^^^^^
+
+Below are some best practices to keep in mind when writing unit tests, in any language.
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst:68]: Word reduction: "Here are some best practices for writing unit tests, in any language."

File: src/chapters/unit-testing/testing-in-java.rst [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+   Every, single, solitary time a test is run, it should produce the same outcome. 
+   A test that passes only most of the time is a worthless test.
+
+#. Relevant
+
+   Tests, as they are written, should be grouped by related class and function.
+
+#. Meaningful
+
+   There is not a need to test trivial code. Unless they contain additional functionality,
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: Typo: "no need".

File: src/chapters/unit-testing/testing-in-java.rst [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+   A test that passes only most of the time is a worthless test.
+
+#. Relevant
+
+   Tests, as they are written, should be grouped by related class and function.
+
+#. Meaningful
+
+   There is not a need to test trivial code. Unless they contain additional functionality, 
+   for example, there is no need to write tests for getters and setters. Your IntelliJ IDE
```
Comment (jimflores5) [inline src/chapters/unit-testing/testing-in-java.rst]: "for example" is extraneous.

File: src/chapters/unit-testing/junit.rst [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+Within ``org.launchcode.java.demos``, open the ``testing`` directory. Inside ``testing``, you'll
+see a ``main`` directory and and ``test`` directory. Open the ``Car`` class within ``main`` and 
+look around. Here, we provide a class ``Car`` with basic information about a make, model, 
+gas level, and mileage as well as getters, setters, and a few other methods. 
+
+In the same directory, you'll find a ``Main`` class with a main method that prints the
+``make`` and ``model`` of a given ``Car`` object. Run this class to verify it works.
+Now, open ``test/CarTest``. It's empty, save for a few TODOs. First, let's tackle the
+first TODO to make a new emptty test. Starting with an empty test lets us validate that we can
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: Typo: "empty"

File: src/chapters/unit-testing/junit.rst [hunk 17]
```diff
@@ -0,0 +1,172 @@
+JUnit
+=====
+
+JUnit is a library that provides the classes, methods, and assertions
+for writing and executing unit tests in Java. In this course, 
+we use JUnit4. Although there are newer versions now available, 
+JUnit4 is still popular and widely used.
+
+Java Annotations
+^^^^^^^^^^^^^^^^
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: The old rst structure works, but to remain consistent with our style guide, use "---" for h2, "^^^" for h3, and "~~~" for h4.

File: src/chapters/unit-testing/junit.rst [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+.. index:: ! annotations
+
+On the topic of unit testing, the annotation ``@Test`` is used to 
+indicate that a public void method should be treated as a test case.
+
+In Java, **annotations** are formalized bits of information about a program. They operate
+somewhere between actual code syntax and a comment on the code. Annotations do not 
+directly affect the code they annotate, but they can supply information to the compiler.
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: "...but they do supply..."

File: src/chapters/unit-testing/junit.rst [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+
+Testing Walkthrough
+^^^^^^^^^^^^^^^^^^^
+
+In your ``java-web-dev-exercises`` repo, we've included two ``.jar`` files for the JUnit 
+library. These are located in the ``lib`` folder. In the future, as you build your own
+Java projects, you will likely include these dependencies in a different fashion. Many Java
+projects use a build automation tool to help manage dependencies, but since this project 
+doesn't use one of these tools, we've included the ``.jar`` files.
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: The discussion of the ``jar`` files breaks the flow of the testing tutorial, especially since there is no explanation of what "dependencies" means.

If we are going to draw attention to the ``jar`` files, then we should spend more time explaining what they store and/or do for us. If knowing about these files is irrelevant for this project, then we can safely ignore them until a later time.

File: src/chapters/unit-testing/junit.rst [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+we use JUnit4. Although there are newer versions now available, 
+JUnit4 is still popular and widely used.
+
+Java Annotations
+^^^^^^^^^^^^^^^^
+
+.. index:: ! annotations
+
+On the topic of unit testing, the annotation ``@Test`` is used to 
+indicate that a public void method should be treated as a test case.
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: Syntax highlighting for ``public void``.

File: src/chapters/unit-testing/junit.rst [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Within ``org.launchcode.java.demos``, open the ``testing`` directory. Inside ``testing``, you'll
+see a ``main`` directory and and ``test`` directory. Open the ``Car`` class within ``main`` and 
+look around. Here, we provide a class ``Car`` with basic information about a make, model, 
+gas level, and mileage as well as getters, setters, and a few other methods. 
+
+In the same directory, you'll find a ``Main`` class with a main method that prints the
+``make`` and ``model`` of a given ``Car`` object. Run this class to verify it works.
+Now, open ``test/CarTest``. It's empty, save for a few TODOs. First, let's tackle the
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: "First...first" correction.  How about, "Let's tackle the first TODO---to make a new, empty test.

File: src/chapters/unit-testing/junit.rst [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+   :linenos:
+
+   //TODO: add emptyTest so we can configure our runtime environment
+   @Test
+   public void emptyTest() {
+      assertEquals(10,10,.001);
+   }
+
+Once written, you likely have one of two states in your IntelliJ environment now. 
+Either IntelliJ sensed which testing library you want to use and added the necessary
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: These two states could be placed into an ``<ol>``.

File: src/chapters/unit-testing/junit.rst [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+   //TODO: add emptyTest so we can configure our runtime environment
+   @Test
+   public void emptyTest() {
+      assertEquals(10,10,.001);
+   }
+
+Once written, you likely have one of two states in your IntelliJ environment now. 
+Either IntelliJ sensed which testing library you want to use and added the necessary 
+import statements at the top of your class file. Or, IntelliJ flagged the ``@Test``
+and ``assertEquals`` lines with red underlines, signalling to you the programmer to
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: Actually, in my IJ version, the words were red, rather than red underlined.

Also, the default corrections suggested by IJ when I hovered over the red words did not import both classes.  Instead, IJ "fixed" the path, imported the Test class and suggested adding the method "assertEquals" to the class.  I had to go to the dropdown to find something similar to the library described in the next code block:

Screen Shot 2019-10-24 at 8 32 29 PM

However, choosing the similar option brought up ANOTHER menu with three more options.  I got the correct one on the second try (it was the middle pick, not the top suggestion).

Long story short - there is a high risk of confusing the students here unless we provide some more guidance.  I'm a fan of learning by doing, but that should be reserved for actual coding rather than figuring out the proper sequence of choices to make with the IDE's helper tools.

How about starting out with something like, "In order to run tests, you will need to import some new classes." Then lead with the import code block and follow with the testing script.

File: src/chapters/unit-testing/junit.rst [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+Java projects, you will likely include these dependencies in a different fashion. Many Java
+projects use a build automation tool to help manage dependencies, but since this project 
+doesn't use one of these tools, we've included the ``.jar`` files.
+
+``main/Car`` and ``test/CarTest``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Within ``org.launchcode.java.demos``, open the ``testing`` directory. Inside ``testing``, you'll
+see a ``main`` directory and and ``test`` directory. Open the ``Car`` class within ``main`` and 
+look around. Here, we provide a class ``Car`` with basic information about a make, model,
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst:53]: Inside the ``Car`` class, there is a ternary conditional operator:
``double milesAbleToTravel = miles > maxDistance ? maxDistance : miles;``

This was never discussed in unit 1, so the syntax needs to be explained or replaced with an ``if/else`` statement.  Searching Google would be tricky, since there is no keyword to guide the query.

File: src/chapters/unit-testing/junit.rst [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+Within ``org.launchcode.java.demos``, open the ``testing`` directory. Inside ``testing``, you'll
+see a ``main`` directory and and ``test`` directory. Open the ``Car`` class within ``main`` and 
+look around. Here, we provide a class ``Car`` with basic information about a make, model, 
+gas level, and mileage as well as getters, setters, and a few other methods. 
+
+In the same directory, you'll find a ``Main`` class with a main method that prints the
+``make`` and ``model`` of a given ``Car`` object. Run this class to verify it works.
+Now, open ``test/CarTest``. It's empty, save for a few TODOs. First, let's tackle the
+first TODO to make a new emptty test. Starting with an empty test lets us validate that we can 
+add JUnit to our classpath, and that the test can be run.
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: We never defined what "classpath" means.

File: src/chapters/unit-testing/junit.rst [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+
+   Car test_car;
+
+   @Before
+   public void createCarObject() {
+      test_car = new Car("Toyota", "Prius", 10, 50);
+   }
+
+Did IntelliJ prompt you to import ``@Before``? Did it import the annotation as you were 
+writing the method? We'll assume you know to check what you need to add to your classpath
```
Comment (jimflores5) [inline src/chapters/unit-testing/junit.rst]: I cannot clearly state why, but the sentence "We'll assume..." splashed some cold water on my mood.  Maybe it's a "you should know this by now" vibe?
